# Shakespeare Text Analysis: NLP Key Phrase Extraction

## Project Overview
This project demonstrates fundamental Natural Language Processing (NLP) techniques applied to classic literature from Shakespeare's plays. The analysis pipeline includes tokenization, stemming, lemmatization, and statistical key phrase extraction using n-gram analysis.

## Objectives
- Extract and analyze text from Shakespeare's plays using the NLTK Gutenberg corpus
- Implement tokenization to break down text into processable units
- Apply stemming and lemmatization for text normalization
- Extract meaningful key phrases using n-gram frequency analysis (monograms, bigrams, and trigrams)
- Identify and contextualize significant phrases within the original text

## Data Source
Shakespeare plays from the Project Gutenberg collection via NLTK's Gutenberg corpus.

## Step 1: Data Loading and Selection
Loading Shakespeare plays from the NLTK Gutenberg corpus. The corpus includes seven plays: Hamlet, Macbeth, Julius Caesar, King Lear, Romeo and Juliet, A Midsummer Night's Dream, and The Merchant of Venice.

In [1]:
import nltk
nltk.download('gutenberg')

from nltk.corpus import gutenberg

plays = [
    "shakespeare-hamlet.txt",
    "shakespeare-macbeth.txt",
    "shakespeare-caesar.txt",
    "shakespeare-lear.txt",
    "shakespeare-romeo_and_juliet.txt",
    "shakespeare-midsummer_night.txt",
    "shakespeare-merchant_of_venice.txt"
]


print("Available Shakespeare Plays (Project Gutenberg):\n")
for i, play in enumerate(plays, start=1):
    print(f"{i}. {play.replace('shakespeare-', '').replace('.txt','').replace('_',' ').title()}")

choice = int(input("\nChoose ONE play (enter number): "))

selected_play = plays[choice - 1]

text = gutenberg.raw(selected_play)

print("\nPlay successfully loaded into variable: text")
print(text[:100])


[nltk_data] Error loading gutenberg: <urlopen error [SSL:
[nltk_data]     CERTIFICATE_VERIFY_FAILED] certificate verify failed:
[nltk_data]     unable to get local issuer certificate (_ssl.c:1000)>


Available Shakespeare Plays (Project Gutenberg):

1. Hamlet
2. Macbeth
3. Caesar
4. Lear
5. Romeo And Juliet
6. Midsummer Night
7. Merchant Of Venice

Play successfully loaded into variable: text
[The Tragedie of Macbeth by William Shakespeare 1603]


Actus Primus. Scoena Prima.

Thunder and Lig


### View Complete Text
Display the full text of the selected play for reference.

In [2]:
# Run to show full printed text
print (text)

[The Tragedie of Macbeth by William Shakespeare 1603]


Actus Primus. Scoena Prima.

Thunder and Lightning. Enter three Witches.

  1. When shall we three meet againe?
In Thunder, Lightning, or in Raine?
  2. When the Hurley-burley's done,
When the Battaile's lost, and wonne

   3. That will be ere the set of Sunne

   1. Where the place?
  2. Vpon the Heath

   3. There to meet with Macbeth

   1. I come, Gray-Malkin

   All. Padock calls anon: faire is foule, and foule is faire,
Houer through the fogge and filthie ayre.

Exeunt.


Scena Secunda.

Alarum within. Enter King Malcome, Donalbaine, Lenox, with
attendants,
meeting a bleeding Captaine.

  King. What bloody man is that? he can report,
As seemeth by his plight, of the Reuolt
The newest state

   Mal. This is the Serieant,
Who like a good and hardie Souldier fought
'Gainst my Captiuitie: Haile braue friend;
Say to the King, the knowledge of the Broyle,
As thou didst leaue it

   Cap. Doubtfull it stood,
As two spent Swimmers, t

## Step 2: Environment Setup
Importing necessary NLP libraries and downloading required NLTK datasets for tokenization, stemming, and lemmatization.

In [3]:
# Run this to import libraries needed
import nltk
from nltk.tokenize import word_tokenize
from nltk.stem import PorterStemmer, WordNetLemmatizer
from nltk.util import ngrams
from collections import Counter
import string

nltk.download('punkt_tab')
nltk.download('punkt')
nltk.download('wordnet')
nltk.download('omw-1.4')

[nltk_data] Error loading punkt_tab: <urlopen error [SSL:
[nltk_data]     CERTIFICATE_VERIFY_FAILED] certificate verify failed:
[nltk_data]     unable to get local issuer certificate (_ssl.c:1000)>
[nltk_data] Error loading punkt: <urlopen error [SSL:
[nltk_data]     CERTIFICATE_VERIFY_FAILED] certificate verify failed:
[nltk_data]     unable to get local issuer certificate (_ssl.c:1000)>
[nltk_data] Error loading wordnet: <urlopen error [SSL:
[nltk_data]     CERTIFICATE_VERIFY_FAILED] certificate verify failed:
[nltk_data]     unable to get local issuer certificate (_ssl.c:1000)>
[nltk_data] Error loading omw-1.4: <urlopen error [SSL:
[nltk_data]     CERTIFICATE_VERIFY_FAILED] certificate verify failed:
[nltk_data]     unable to get local issuer certificate (_ssl.c:1000)>


False

## Step 3: Tokenization
Breaking down the text into individual tokens (words and punctuation). Tokenization is the foundational step for all subsequent NLP analysis, converting raw text into structured data.

In [4]:
# Place code for Tokenization below here, do not delete this comment
tokens = word_tokenize(text)

print("Tokens:")
print(tokens)

Tokens:
['[', 'The', 'Tragedie', 'of', 'Macbeth', 'by', 'William', 'Shakespeare', '1603', ']', 'Actus', 'Primus', '.', 'Scoena', 'Prima', '.', 'Thunder', 'and', 'Lightning', '.', 'Enter', 'three', 'Witches', '.', '1', '.', 'When', 'shall', 'we', 'three', 'meet', 'againe', '?', 'In', 'Thunder', ',', 'Lightning', ',', 'or', 'in', 'Raine', '?', '2', '.', 'When', 'the', 'Hurley-burley', "'s", 'done', ',', 'When', 'the', 'Battaile', "'s", 'lost', ',', 'and', 'wonne', '3', '.', 'That', 'will', 'be', 'ere', 'the', 'set', 'of', 'Sunne', '1', '.', 'Where', 'the', 'place', '?', '2', '.', 'Vpon', 'the', 'Heath', '3', '.', 'There', 'to', 'meet', 'with', 'Macbeth', '1', '.', 'I', 'come', ',', 'Gray-Malkin', 'All', '.', 'Padock', 'calls', 'anon', ':', 'faire', 'is', 'foule', ',', 'and', 'foule', 'is', 'faire', ',', 'Houer', 'through', 'the', 'fogge', 'and', 'filthie', 'ayre', '.', 'Exeunt', '.', 'Scena', 'Secunda', '.', 'Alarum', 'within', '.', 'Enter', 'King', 'Malcome', ',', 'Donalbaine', ',', 'Le

## Step 4: Stemming
Applying the Porter Stemmer algorithm to reduce words to their root forms. Stemming uses heuristic rules to chop off word endings, which may result in non-dictionary words but ensures consistent grouping of related terms.

In [5]:
# Place code for Stemming below here, do not delete this comment
stemmer = PorterStemmer()
stems = [stemmer.stem(token) for token in tokens]
print("\nStems:")
print(stems)


Stems:
['[', 'the', 'tragedi', 'of', 'macbeth', 'by', 'william', 'shakespear', '1603', ']', 'actu', 'primu', '.', 'scoena', 'prima', '.', 'thunder', 'and', 'lightn', '.', 'enter', 'three', 'witch', '.', '1', '.', 'when', 'shall', 'we', 'three', 'meet', 'again', '?', 'in', 'thunder', ',', 'lightn', ',', 'or', 'in', 'rain', '?', '2', '.', 'when', 'the', 'hurley-burley', "'s", 'done', ',', 'when', 'the', 'battail', "'s", 'lost', ',', 'and', 'wonn', '3', '.', 'that', 'will', 'be', 'ere', 'the', 'set', 'of', 'sunn', '1', '.', 'where', 'the', 'place', '?', '2', '.', 'vpon', 'the', 'heath', '3', '.', 'there', 'to', 'meet', 'with', 'macbeth', '1', '.', 'i', 'come', ',', 'gray-malkin', 'all', '.', 'padock', 'call', 'anon', ':', 'fair', 'is', 'foul', ',', 'and', 'foul', 'is', 'fair', ',', 'houer', 'through', 'the', 'fogg', 'and', 'filthi', 'ayr', '.', 'exeunt', '.', 'scena', 'secunda', '.', 'alarum', 'within', '.', 'enter', 'king', 'malcom', ',', 'donalbain', ',', 'lenox', ',', 'with', 'attend'

## Step 5: Lemmatization
Using WordNet Lemmatizer to reduce words to their base dictionary forms. Unlike stemming, lemmatization uses vocabulary and morphological analysis to return actual dictionary words, providing more linguistically accurate normalization.

In [6]:
# Place code for Lemmatization below here, do not delete this comment
lemmatizer = WordNetLemmatizer()
lemmas = [lemmatizer.lemmatize(token) for token in tokens]
print("\nLemmas:")
print(lemmas)


Lemmas:
['[', 'The', 'Tragedie', 'of', 'Macbeth', 'by', 'William', 'Shakespeare', '1603', ']', 'Actus', 'Primus', '.', 'Scoena', 'Prima', '.', 'Thunder', 'and', 'Lightning', '.', 'Enter', 'three', 'Witches', '.', '1', '.', 'When', 'shall', 'we', 'three', 'meet', 'againe', '?', 'In', 'Thunder', ',', 'Lightning', ',', 'or', 'in', 'Raine', '?', '2', '.', 'When', 'the', 'Hurley-burley', "'s", 'done', ',', 'When', 'the', 'Battaile', "'s", 'lost', ',', 'and', 'wonne', '3', '.', 'That', 'will', 'be', 'ere', 'the', 'set', 'of', 'Sunne', '1', '.', 'Where', 'the', 'place', '?', '2', '.', 'Vpon', 'the', 'Heath', '3', '.', 'There', 'to', 'meet', 'with', 'Macbeth', '1', '.', 'I', 'come', ',', 'Gray-Malkin', 'All', '.', 'Padock', 'call', 'anon', ':', 'faire', 'is', 'foule', ',', 'and', 'foule', 'is', 'faire', ',', 'Houer', 'through', 'the', 'fogge', 'and', 'filthie', 'ayre', '.', 'Exeunt', '.', 'Scena', 'Secunda', '.', 'Alarum', 'within', '.', 'Enter', 'King', 'Malcome', ',', 'Donalbaine', ',', 'Le

## Step 6: Key Phrase Extraction Using N-Gram Analysis
Extracting candidate key phrases through statistical frequency analysis:
- **Monograms (1-grams)**: Single words that appear frequently
- **Bigrams (2-grams)**: Two-word sequences that may represent important concepts
- **Trigrams (3-grams)**: Three-word sequences that capture longer phrases

The extraction process involves:
1. Text preprocessing (lowercasing, punctuation removal)
2. Lemmatization for normalization
3. N-gram generation
4. Frequency counting to identify the most significant phrases

In [7]:
# Place code for Key Phrase Extraction below here, do not delete this comment

# Clean tokens (lowercase, remove punctuation)
clean_tokens = [token.lower() for token in tokens if token not in string.punctuation]

# Lemmatize clean tokens
lemmas = [lemmatizer.lemmatize(token) for token in clean_tokens]

# Monograms (1-gram)
monograms = list(ngrams(lemmas, 1))

# Bigrams (2-gram)
bigrams = list(ngrams(lemmas, 2))

# Trigrams (3-gram)
trigrams = list(ngrams(lemmas, 3))

# Count frequencies
monogram_freq = Counter(monograms)
bigram_freq = Counter(bigrams)
trigram_freq = Counter(trigrams)

print("Candidate Keyphrases (Monograms):")
for phrase, freq in monogram_freq.most_common(10):
    print(" ".join(phrase), "→", freq)

print("\nCandidate Keyphrases (Bigrams):")
for phrase, freq in bigram_freq.most_common(10):
    print(" ".join(phrase), "→", freq)

print("\nCandidate Keyphrases (Trigrams):")
for phrase, freq in trigram_freq.most_common(10):
    print(" ".join(phrase), "→", freq)

Candidate Keyphrases (Monograms):
the → 649
and → 545
to → 383
of → 338
i → 331
a → 328
that → 236
'd → 224
is → 216
you → 205

Candidate Keyphrases (Bigrams):
i haue → 41
't is → 36
i am → 29
of the → 27
to the → 27
in the → 26
and the → 22
my lord → 22
can not → 20
to be → 19

Candidate Keyphrases (Trigrams):
thane of cawdor → 13
the thane of → 8
my good lord → 8
i pray you → 7
can not be → 6
who 's there → 6
knock knock knock → 6
i can not → 5
enter macbeth macb → 5
what 's the → 5


## Analysis Results and Insights

### 1. Vocabulary Size Analysis
Calculating the number of unique tokens to understand the lexical diversity of the text.

In [8]:
# 1) How many unique tokens were you able to extract?
unique_token_count = len(set(tokens))
print(f"Number of unique tokens extracted: {unique_token_count}")

Number of unique tokens extracted: 4076


### 2. Top Key Phrases by Category
Identifying the most frequent phrases across all n-gram categories to understand the text's dominant themes and linguistic patterns.

**Monograms (Single Words):**

In [9]:
# 2) What are the top 10 Key Phrases in your assigned play

print("Top 10 Candidate Keyphrases (Monograms):")
for phrase, freq in monogram_freq.most_common(10):
    print(" ".join(phrase), "→", freq)

Top 10 Candidate Keyphrases (Monograms):
the → 649
and → 545
to → 383
of → 338
i → 331
a → 328
that → 236
'd → 224
is → 216
you → 205


**Bigrams (Two-Word Phrases):**

In [10]:
print("\nTop 10 Candidate Keyphrases (Bigrams):")
for phrase, freq in bigram_freq.most_common(10):
    print(" ".join(phrase), "→", freq)



Top 10 Candidate Keyphrases (Bigrams):
i haue → 41
't is → 36
i am → 29
of the → 27
to the → 27
in the → 26
and the → 22
my lord → 22
can not → 20
to be → 19


**Trigrams (Three-Word Phrases):**

In [11]:
print("\nTop 10 Candidate Keyphrases (Trigrams):")
for phrase, freq in trigram_freq.most_common(10):
    print(" ".join(phrase), "→", freq)


Top 10 Candidate Keyphrases (Trigrams):
thane of cawdor → 13
the thane of → 8
my good lord → 8
i pray you → 7
can not be → 6
who 's there → 6
knock knock knock → 6
i can not → 5
enter macbeth macb → 5
what 's the → 5


### 3. Contextual Analysis of Key Phrases
Locating high-frequency key phrases within the original text to verify their significance and understand their usage context. This helps validate that our extracted phrases are truly meaningful in the context of the play.

In [12]:
# 3) Choose one key phrase from your output and locate it in the original text.
import re

# Choose a meaningful keyphrase (e.g., top bigram: "my lord")
chosen_phrase = "my lord"

print(f"Searching for key phrase: '{chosen_phrase}'\n")

pattern = re.compile(re.escape(chosen_phrase), re.IGNORECASE)
matches = list(pattern.finditer(text))

print(f"Found {len(matches)} occurrences in the original text.\n")

for i, match in enumerate(matches[:3], 1):
    start = max(0, match.start() - 50)
    end = min(len(text), match.end() + 50)
    context = text[start:end].replace('\n', ' ')
    print(f"Occurrence {i}: ...{context}...")

Searching for key phrase: 'my lord'

Found 22 occurrences in the original text.

Occurrence 1: ... to morrow. Is't farre you ride?   Ban. As farre, my Lord, as will fill vp the time 'Twixt this, and Supper...
Occurrence 2: ... or twaine     Macb. Faile not our Feast     Ban. My Lord, I will not     Macb. We heare our bloody Cozens ...
Occurrence 3: ...tend those men Our pleasure?   Seruant. They are, my Lord, without the Pallace Gate     Macb. Bring them be...


### 4. N-Gram Distribution Summary
Analyzing the distribution of different n-gram types in the top key phrases to understand which phrase lengths are most significant in the text.

In [13]:
# 4) How many monograms, bigrams, trigrams, and n-grams appeared in the top 10 keyphrases

# Count from each category's top 10
mono_top10 = monogram_freq.most_common(10)
bi_top10 = bigram_freq.most_common(10)
tri_top10 = trigram_freq.most_common(10)

print("Top 10 from each n-gram category:")
print(f"  - Monograms (1-word): {len(mono_top10)}")
print(f"  - Bigrams (2-word): {len(bi_top10)}")
print(f"  - Trigrams (3-word): {len(tri_top10)}")
print(f"\nTotal keyphrases analyzed: {len(mono_top10) + len(bi_top10) + len(tri_top10)}")

Top 10 from each n-gram category:
  - Monograms (1-word): 10
  - Bigrams (2-word): 10
  - Trigrams (3-word): 10

Total keyphrases analyzed: 30


---

## Key Findings from Macbeth Analysis

### Vocabulary Insights
- **Total Unique Tokens**: 4,076 distinct words
- This relatively rich vocabulary reflects Shakespeare's complex linguistic style and the play's exploration of multiple themes

### Dominant Language Patterns
The n-gram analysis revealed several important linguistic characteristics of Macbeth:

**Most Significant Bigrams:**
- "i haue" (41 occurrences) - Old English spelling patterns
- "'t is" (36 occurrences) - Common Shakespearean contraction
- "my lord" (22 occurrences) - Formal address reflecting the play's focus on nobility and hierarchy

**Most Significant Trigrams:**
- "thane of cawdor" (13 occurrences) - Central to the plot and Macbeth's ambition
- "knock knock knock" (6 occurrences) - The famous porter scene, a pivotal moment of dramatic tension
- "my good lord" (8 occurrences) - Formality and social structure

### Literary and Thematic Observations
1. **Hierarchical Society**: Phrases like "my lord" and "thane of cawdor" dominate, showing the importance of rank and title
2. **Character Relationships**: High frequency of "i haue" and "i am" suggests introspective dialogue and personal agency
3. **Dramatic Tension**: The repetition of "knock knock knock" captures a key atmospheric moment in the play
4. **Old English Orthography**: Patterns like "'d", "'t is", and "haue" preserve Shakespeare's original spellings

## Conclusion: What I Learned About Analyzing Macbeth

### Insights About Macbeth's Linguistic Landscape
Through this NLP analysis of Shakespeare's Macbeth, I discovered several fascinating patterns:

**1. The Language of Power and Hierarchy**
The play's vocabulary is dominated by formal titles and addresses. The phrase "my lord" appears 22 times, and "thane of cawdor" - central to Macbeth's fatal ambition - appears 13 times. This quantitative evidence confirms that Macbeth is fundamentally a play about power, ambition, and social rank. The frequent use of formal address reflects the rigid hierarchical structure of medieval Scottish society.

**2. Character Voice and Agency**
The high frequency of "i haue" (41 times) and "i am" (29 times) reveals that Macbeth features characters who assert their agency and actions. This is fitting for a tragedy driven by personal choice and moral corruption - characters constantly declare what they have done or who they are.

**3. Dramatic Moments Captured Statistically**
The trigram "knock knock knock" appearing 6 times might seem insignificant, but it represents one of the play's most famous scenes - the porter's scene after Duncan's murder. This shows how NLP can identify dramatically significant moments through repetition.

**4. Lexical Richness**
With 4,076 unique tokens, Macbeth demonstrates Shakespeare's linguistic complexity. The vocabulary includes archaic spellings (haue, 't is) that preserve the historical authenticity of the text.

### Technical Insights Gained
- **Lemmatization vs. Raw Text**: By lemmatizing tokens, I was able to group related word forms and identify true thematic patterns rather than just counting spelling variations
- **N-gram Effectiveness**: Bigrams and trigrams proved more meaningful than monograms for understanding content - while monograms were mostly function words ("the", "and"), bigrams captured actual phrases like "my lord"
- **Context Validation**: Locating key phrases in the original text confirmed that frequency-based extraction does identify genuinely important elements

### Skills Demonstrated
This project showcases my ability to:
- Apply NLP preprocessing techniques (tokenization, stemming, lemmatization) to literary texts
- Extract and interpret statistical patterns from unstructured text data
- Connect quantitative analysis to qualitative literary interpretation
- Use Python libraries (NLTK, Collections) for text analytics
- Document technical work in a clear, professional manner

### Future Applications
These NLP techniques could extend to:
- Comparative analysis across multiple Shakespeare plays
- Character-specific language pattern analysis
- Sentiment analysis of different acts or scenes
- Author attribution studies
- Automated theme identification

### Technologies Used
- **Python 3.12.8**
- **NLTK (Natural Language Toolkit)** for text processing
- **Collections.Counter** for frequency analysis
- **Regular Expressions** for pattern matching
- **Jupyter Notebook** for interactive analysis